irksome/stage.py

```diff
@@ -252,85 +275,113 @@ def getFormStage(F, butch, u0, t, dt, bcs=None, splitting=None,
-                repl[u0bits[k][ii]] = UUbits[i][k][ii]
+    nspacenew = getNullspace(V, Vbig, butch, nullspace)
 
-        eFFi = replace(split_form.remainder, repl)
+    # only form update stuff if we need it
+    # which means neither stiffly accurate nor Vandermonde
+    if not butch.is_stiffly_accurate:
+        unew = Function(V)
 
-        Fupdate += dt * B[i] * eFFi
+        Fupdate = inner(unew - u0, v) * dx
+        B = vectorize(lambda c: MC.Constant(c))(butch.b)
+        C = vectorize(lambda c: MC.Constant(c))(butch.c)
 
-    # And the BC's for the update -- just the original BC at t+dt
-    update_bcs = []
-    update_bcs_gblah = []
-    for bc in bcs:
-        if num_fields == 1:  # not mixed space
-            comp = bc.function_space().component
-            if comp is not None:  # check for sub-piece of vector-valued
-                Vsp = V.sub(comp)
-            else:
-                Vsp = V
-        else:  # mixed space
-            sub = bc.function_space_index()
-            comp = bc.function_space().component
-            if comp is not None:  # check for sub-piece of vector-valued
-                Vsp = V.sub(sub).sub(comp)
-            else:
-                Vsp = V.sub(sub)
+        for i in range(num_stages):
+            repl = {t: t + C[i] * dt}
 
-        bcarg = as_ufl(bc._original_arg)
-        try:
-            gdat = assemble(interpolate(bcarg, Vsp))
-            gmethod = lambda gd, gc: gd.interpolate(gc)
-        except:  # noqa: E722
-            gdat = project(bcarg, Vsp)
-            gmethod = lambda gd, gc: gd.project(gc)
+            for k in range(num_fields):
+                repl[u0bits[k]] = UUbits[i][k]
+                for ii in np.ndindex(u0bits[k].ufl_shape):
+                    repl[u0bits[k][ii]] = UUbits[i][k][ii]
+
+            eFFi = replace(split_form.remainder, repl)
+
+            Fupdate += dt * B[i] * eFFi
+
+        # And the BC's for the update -- just the original BC at t+dt
+        update_bcs = []
+        update_bcs_gblah = []
+        for bc in bcs:
+            if num_fields == 1:  # not mixed space
+                comp = bc.function_space().component
+                if comp is not None:  # check for sub-piece of vector-valued
+                    Vsp = V.sub(comp)
+                else:
+                    Vsp = V
+            else:  # mixed space
+                sub = bc.function_space_index()
+                comp = bc.function_space().component
+                if comp is not None:  # check for sub-piece of vector-valued
+                    Vsp = V.sub(sub).sub(comp)
+                else:
+                    Vsp = V.sub(sub)
+
+            bcarg = as_ufl(bc._original_arg)
+            try:
+                gdat = assemble(interpolate(bcarg, Vsp))
+                gmethod = lambda gd, gc: gd.interpolate(gc)
+            except:  # noqa: E722
+                gdat = project(bcarg, Vsp)
+                gmethod = lambda gd, gc: gd.project(gc)
```

Comment:
As in the demo, I think we need some solver options for this projection in the case where we want to satisfy bounds constraints.

Reply:
Extra logic is added.